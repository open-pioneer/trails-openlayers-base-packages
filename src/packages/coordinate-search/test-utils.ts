// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
