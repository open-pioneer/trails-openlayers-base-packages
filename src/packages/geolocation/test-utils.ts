// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
