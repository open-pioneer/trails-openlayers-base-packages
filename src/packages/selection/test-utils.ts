// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Reactive, reactive } from "@conterra/reactivity-core";
import {
    SelectionSource,
    SelectionResult,
    SelectionOptions,
    SelectionSourceStatus,
    SelectionKind
} from "./api";
import { Point } from "ol/geom";

export const fakeSelectedPointFeatures = [
    new Point([407354, 5754673]), // con terra (Bottom Right)
    new Point([404740, 5757893]) // Schloss (Top Left)
];

let count = 0;

export class FakePointSelectionSource implements SelectionSource {
    readonly label = `Fake Selection Source #${++count}`;
    #timeout: number;
    #status: Reactive<SelectionSourceStatus>;
    #pointFeatures: Point[];

    constructor(
        timeout?: number,
        status?: SelectionSourceStatus,
        pointFeatures: Point[] = fakeSelectedPointFeatures
    ) {
        this.#timeout = timeout || 0;
        this.#status = reactive(status || "unavailable");
        this.#pointFeatures = pointFeatures;
    }

    get status(): SelectionSourceStatus {
        return this.#status.value;
    }

    set status(value: SelectionSourceStatus) {
        this.#status.value = value;
    }

    async select(
        selectionKind: SelectionKind,
        options: SelectionOptions
    ): Promise<SelectionResult[]> {
        if (selectionKind.type !== "extent") {
            throw new Error(`Unsupported selection kind: ${selectionKind.type}`);
        }

        if (this.#status.value !== "available") return [];

        await new Promise((resolve) => setTimeout(resolve, this.#timeout));

        const allPoints = this.#pointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            if (!point.intersectsExtent(selectionKind.extent)) {
                return undefined;
            }
            return result;
        });

        const selectedPoints = allPoints.filter((s): s is SelectionResult => s != null);
        const limitedResults =
            selectedPoints.length > options.maxResults
                ? selectedPoints.slice(0, options.maxResults)
                : selectedPoints;
        return limitedResults;
    }
}

/**
 * For testing purposes only
 */
export class NoStatusSelectionSource implements SelectionSource {
    readonly label: string = "Testlabel";

    async select(_: SelectionKind, __: SelectionOptions): Promise<SelectionResult[]> {
        const allPoints = fakeSelectedPointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            return result;
        });
        return allPoints;
    }
}
