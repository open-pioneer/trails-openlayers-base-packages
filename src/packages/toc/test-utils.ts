// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { screen } from "@testing-library/react";

export async function findToc() {
    const tocDiv = await screen.findByTestId("toc");
    return tocDiv;
}
