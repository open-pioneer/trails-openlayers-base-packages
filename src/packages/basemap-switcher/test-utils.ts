// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
