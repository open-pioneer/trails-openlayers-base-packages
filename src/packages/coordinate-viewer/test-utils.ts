// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
