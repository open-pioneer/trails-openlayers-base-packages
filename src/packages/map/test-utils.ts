// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { screen, waitFor } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import {
    ExtentConfig,
    InitialViewConfig,
    LayerConfig,
    MapConfig,
    MapConfigProvider,
    MapRegistry,
    OlMapOptions
} from "./api";
import { MapRegistryImpl } from "./services";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
    extent?: ExtentConfig;
    projection?: string;
    layers?: LayerConfig[];
    advanced?: OlMapOptions;

    noInitialView?: boolean;
    noProjection?: boolean;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const getInitialView = (): InitialViewConfig => {
        if (options?.extent) {
            return {
                kind: "extent",
                extent: options.extent
            };
        }
        return {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        };
    };

    const mapConfig: MapConfig = {
        initialView: options?.noInitialView ? undefined : getInitialView(),
        projection: options?.noProjection ? undefined : options?.projection ?? "EPSG:3857",
        layers: options?.layers ?? [
            {
                title: "OSM",
                layer: new TileLayer({
                    source: new OSM()
                })
            }
        ],
        advanced: options?.advanced
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
