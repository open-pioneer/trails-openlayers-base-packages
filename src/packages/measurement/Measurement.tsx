// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    Box,
    Button,
    FormControl,
    FormLabel,
    HStack,
    Select,
    Text
} from "@open-pioneer/chakra-integration";
import { MapModel, MapModelProps, useMapModel } from "@open-pioneer/map";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { PackageIntl } from "@open-pioneer/runtime";
import { Circle as CircleStyle, Fill, Stroke, Style } from "ol/style";
import { StyleLike } from "ol/style/Style";
import { useIntl } from "open-pioneer:react-hooks";
import { FC, useEffect, useState } from "react";
import { MeasurementController, MeasurementType } from "./MeasurementController";
import LineString from "ol/geom/LineString";
import Polygon from "ol/geom/Polygon";

/** Emitted when a new measurement is being added to the map. */
export interface MeasurementsAddEvent {
    kind: "add-measurement";
    geometry: MeasurementGeometry;
}

/** Emitted when a measurement is being removed from the map. */
export interface MeasurementsRemoveEvent {
    kind: "remove-measurement";
    geometry: MeasurementGeometry;
}

/**
 * A change event emitted by {@link Measurement} when measurements change.
 *
 * NOTE: Non-exhaustive. More event types may be added in a future version.
 **/
export type MeasurementsChangeEvent = MeasurementsAddEvent | MeasurementsRemoveEvent;

/**
 * Represents the geometry of a measurement.
 */
export type MeasurementGeometry = LineString | Polygon;

/**
 * This is for special properties of the Measurement.
 */
export interface MeasurementProps extends CommonComponentProps, MapModelProps {
    /**
     * The style for the active drawing feature's geometry.
     */
    activeFeatureStyle?: StyleLike;

    /**
     * The style for the finished drawn feature's geometry.
     */
    finishedFeatureStyle?: StyleLike;

    /**
     * List of measurements to be rendered when the component is initialized.
     */
    predefinedMeasurements?: MeasurementGeometry[];

    /**
     * Event handler that is called whenever a measurement is added or removed.
     */
    onMeasurementsChange?: (event: MeasurementsChangeEvent) => void;
}

/**
 * The `Measurement` component can be used in an app to measure geometries (areas or lines) on the map.
 */
export const Measurement: FC<MeasurementProps> = (props) => {
    const intl = useIntl();
    const { containerProps } = useCommonComponentProps("measurement", props);
    const [selectedMeasurement, setMeasurement] = useState<MeasurementType>("distance");
    const label = (id: string) => intl.formatMessage({ id: id });
    const mapState = useMapModel(props);
    const controller = useController(mapState.map, props, intl);

    // Start / Stop measurement on selection change
    useEffect(() => {
        if (!controller) {
            return;
        }

        controller.startMeasurement(selectedMeasurement);
        return () => {
            controller.stopMeasurement();
        };
    }, [controller, selectedMeasurement]);

    function changeMeasurement(measurementType: string) {
        if (measurementType === "distance" || measurementType === "area") {
            setMeasurement(measurementType);
        } else {
            throw new Error(`Unexpected measurement type: '${measurementType}'.`);
        }
    }

    function removeMeasurements() {
        controller?.clearMeasurements();
    }

    return (
        <Box {...containerProps}>
            <Text mb={3}>{intl.formatMessage({ id: "measurementInfoText" })}</Text>
            <FormControl mb={4} alignItems="center">
                <HStack mb={2}>
                    <FormLabel mb={1}>{label("measurementLabel")}</FormLabel>
                    <Select
                        value={selectedMeasurement}
                        onChange={(e) => changeMeasurement(e.target.value)}
                        className="measurement-select"
                    >
                        <option value={"distance"}>{label("distance")}</option>
                        <option value={"area"}>{label("area")}</option>
                    </Select>
                </HStack>
            </FormControl>
            <Button
                padding={2}
                className="measurement-delete-button"
                onClick={removeMeasurements}
                width="100%"
            >
                {label("deleteMeasurementLabel")}
            </Button>
        </Box>
    );
};

/** Creates a MeasurementController instance for the given map. */
function useController(map: MapModel | undefined, props: MeasurementProps, intl: PackageIntl) {
    const {
        activeFeatureStyle,
        finishedFeatureStyle,
        onMeasurementsChange,
        predefinedMeasurements
    } = props;
    const [controller, setController] = useState<MeasurementController | undefined>(undefined);
    useEffect(() => {
        if (!map) {
            return;
        }
        const controller = new MeasurementController(map.olMap, {
            getContinueMessage() {
                return intl.formatMessage({ id: "tooltips.continue" });
            },
            getHelpMessage() {
                return intl.formatMessage({ id: "tooltips.help" });
            },
            formatNumber(value) {
                return intl.formatNumber(value, {
                    maximumFractionDigits: 2
                });
            }
        });

        setController(controller);
        return () => {
            controller.destroy();
            setController(undefined);
        };
    }, [map, intl]);

    // Synchronize styles with controller
    useEffect(() => {
        controller?.setActiveFeatureStyle(activeFeatureStyle ?? getDefaultActiveFeatureStyle());
    }, [controller, activeFeatureStyle]);
    useEffect(() => {
        controller?.setFinishedFeatureStyle(
            finishedFeatureStyle ?? getDefaultFinishedFeatureStyle()
        );
    }, [controller, finishedFeatureStyle]);
    useEffect(() => {
        controller?.setMeasurementSourceChangedHandler(onMeasurementsChange);
    }, [controller, onMeasurementsChange]);

    useEffect(() => {
        controller?.setPredefinedMeasurements(predefinedMeasurements ?? []);
    }, [controller, predefinedMeasurements]);
    return controller;
}

function getDefaultActiveFeatureStyle() {
    return [
        new Style({
            stroke: new Stroke({
                color: "#fff",
                lineDash: [10, 10],
                width: 5
            })
        }),
        new Style({
            fill: new Fill({
                color: "rgba(0,0,0,0.15)"
            }),
            stroke: new Stroke({
                color: "rgba(0, 0, 0, 0.7)",
                lineDash: [10, 10],
                width: 3
            }),
            image: new CircleStyle({
                radius: 5,
                stroke: new Stroke({
                    color: "rgba(0, 0, 0, 0.7)",
                    width: 2
                }),
                fill: new Fill({
                    color: "rgba(255, 255, 255, 0.2)"
                })
            })
        })
    ];
}

function getDefaultFinishedFeatureStyle() {
    return [
        new Style({
            stroke: new Stroke({
                color: "#fff",
                width: 5
            })
        }),
        new Style({
            stroke: new Stroke({
                color: "#0b96fc",
                width: 3
            }),
            fill: new Fill({
                color: "rgba(11,150,252,0.15)"
            })
        })
    ];
}
