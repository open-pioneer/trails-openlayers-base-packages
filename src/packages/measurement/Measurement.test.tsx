// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0

import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { Measurement } from "./Measurement";
import { expect, it } from "vitest";

it("should successfully create a measurement component", async () => {
    render(
        <PackageContextProvider>
            <Measurement data-testid="measurement"></Measurement>
        </PackageContextProvider>
    );

    // measurement is mounted
    const { measurementDiv, measurementSelectDiv, measurementSelect } = await waitForMeasurement();
    expect(measurementDiv).toMatchSnapshot();

    // check measurement select is available
    expect(measurementSelectDiv).toBeInstanceOf(HTMLDivElement);
    expect(measurementSelect).toBeInstanceOf(HTMLSelectElement);
});

it("should successfully create a measurement component with additional css classes and box properties", async () => {
    render(
        <PackageContextProvider>
            <Measurement className="test" data-testid="measurement"></Measurement>
        </PackageContextProvider>
    );

    // measurement is mounted
    const { measurementDiv } = await waitForMeasurement();
    expect(measurementDiv).toMatchSnapshot();

    expect(measurementDiv.classList.contains("test")).toBe(true);
    expect(measurementDiv.classList.contains("foo")).toBe(false);
});

it("should successfully select a measurement from the select dropdown", async () => {
    render(
        <PackageContextProvider>
            <Measurement className="test" data-testid="measurement"></Measurement>
        </PackageContextProvider>
    );

    // measurement is mounted
    const { measurementSelect } = await waitForMeasurement();

    act(() => {
        fireEvent.change(measurementSelect, { target: { value: "area" } });
    });
    expect(measurementSelect.value).toBe("area");

    act(() => {
        fireEvent.change(measurementSelect, { target: { value: "distance" } });
    });
    expect(measurementSelect.value).toBe("distance");
});

async function waitForMeasurement() {
    const { measurementDiv, measurementSelectDiv, measurementSelect } = await waitFor(async () => {
        const measurementDiv: HTMLDivElement | null =
            await screen.findByTestId<HTMLDivElement>("measurement");
        if (!measurementDiv) {
            throw new Error("Measurement not rendered");
        }

        const measurementSelectDiv = measurementDiv.querySelector(".measurement-content");
        if (!measurementSelectDiv) {
            throw new Error("Measurement select container not rendered");
        }

        const measurementSelect =
            measurementSelectDiv.querySelector<HTMLSelectElement>(".measurement-select");

        if (!measurementSelect) {
            throw new Error("Measurement select not rendered");
        }

        return { measurementDiv, measurementSelectDiv, measurementSelect };
    });

    return { measurementDiv, measurementSelectDiv, measurementSelect };
}
