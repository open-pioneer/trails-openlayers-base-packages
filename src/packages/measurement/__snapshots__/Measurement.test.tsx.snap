// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a measurement component 1`] = `
<div
  class="measurement css-0"
  data-testid="measurement"
  data-theme="light"
>
  <p
    class="chakra-text css-93tr2s"
    data-theme="light"
  >
    measurementInfoText
  </p>
  <div
    class="chakra-form-control css-1m7igkt"
    data-theme="light"
    role="group"
  >
    <div
      class="chakra-stack css-1sgivba"
      data-theme="light"
    >
      <label
        class="chakra-form__label css-oe420r"
        data-theme="light"
        for="field-:r0:"
        id="field-:r0:-label"
      >
        measurementLabel
      </label>
      <div
        class="chakra-select__wrapper css-42b2qy"
        data-theme="light"
      >
        <select
          class="chakra-select measurement-select css-161pkch"
          data-theme="light"
          id="field-:r0:"
        >
          <option
            value="distance"
          >
            distance
          </option>
          <option
            value="area"
          >
            area
          </option>
        </select>
        <div
          class="chakra-select__icon-wrapper css-iohxn1"
          data-theme="light"
        >
          <svg
            aria-hidden="true"
            class="chakra-select__icon"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentcolor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <button
    class="chakra-button measurement-delete-button css-14p1tvq"
    data-theme="light"
    type="button"
  >
    deleteMeasurementLabel
  </button>
</div>
`;
